index/server/pkg/server/endpoint.go

```diff
@@ -0,0 +1,262 @@
+package server
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"net/http/httputil"
+	"net/url"
+	"os"
+	"path"
+
+	indexSchema "github.com/devfile/registry-support/index/generator/schema"
+	"github.com/devfile/registry-support/index/server/pkg/util"
+	"github.com/gin-gonic/gin"
+	"github.com/prometheus/client_golang/prometheus"
+	"gopkg.in/segmentio/analytics-go.v3"
+)
+
+// queryParams is a struct that contains a list of query options for the registry API
+type queryParams struct {
+	indexType     string
+	iconType      string
+	architectures []string
+}
+
+// serveRootEndpoint sets up the handler for the root (/) endpoint on the server
+// If html is requested (i.e. from a web browser), the viewer is displayed, otherwise the devfile index is served.
+func serveRootEndpoint(c *gin.Context) {
+	// Determine if text/html was requested by the client
+	acceptHeader := c.Request.Header.Values("Accept")
+	if util.IsHtmlRequested(acceptHeader) {
+		c.Redirect(http.StatusFound, "/viewer")
+	} else {
+		serveDevfileIndex(c)
+	}
+}
+
+// serveDevfileIndex serves the index.json file located in the container at `serveDevfileIndex`
+func serveDevfileIndex(c *gin.Context) {
+	// Start the counter for the request
+	var status string
+	timer := prometheus.NewTimer(prometheus.ObserverFunc(func(v float64) {
+		getIndexLatency.WithLabelValues(status).Observe(v)
+	}))
+	defer func() {
+		timer.ObserveDuration()
+	}()
+
+	params := queryParams{
+		indexType:     "stack",
+		iconType:      c.Query("icon"),
+		architectures: c.QueryArray("architectures"),
+	}
+
+	// Serve the index.json file
+	buildIndexAPIResponse(c, params)
+}
+
+// serveDevfileIndexWithType returns the index file content with specific devfile type
+func serveDevfileIndexWithType(c *gin.Context) {
+
+	params := queryParams{
```

Comment:
I wonder if it would make sense to throw `params` into `buildIndexAPIResponse` rather than initializing the same set of parameters twice? 

Basically remove `params` from `serveDevfileIndex` and `serveDevfileIndexWithType` and move it into `buildIndexAPIResponse`?

Reply:
make sense